orderbook/src/api/validation.rs

```diff
@@ -0,0 +1,353 @@
+use contracts::WETH9;
+use ethcontract::H160;
+use model::order::{BuyTokenDestination, Order, SellTokenSource, BUY_ETH_ADDRESS};
+use shared::web3_traits::CodeFetching;
+use std::time::Duration;
+use warp::{http::StatusCode, reply::Json};
+
+#[derive(Debug)]
+pub enum ValidationError {
+    Forbidden,
+    InsufficientValidTo,
+    TransferEthToContract,
+    SameBuyAndSellToken,
+    UnsupportedBuyTokenDestination(BuyTokenDestination),
+    UnsupportedSellTokenSource(SellTokenSource),
+    Other(anyhow::Error),
+}
+
+impl ValidationError {
+    pub fn to_warp_reply(&self) -> (Json, StatusCode) {
+        match self {
+            Self::UnsupportedBuyTokenDestination(dest) => (
+                super::error("UnsupportedBuyTokenDestination", format!("Type {:?}", dest)),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::UnsupportedSellTokenSource(src) => (
+                super::error("UnsupportedSellTokenSource", format!("Type {:?}", src)),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::Forbidden => (
+                super::error("Forbidden", "Forbidden, your account is deny-listed"),
+                StatusCode::FORBIDDEN,
+            ),
+            Self::InsufficientValidTo => (
+                super::error(
+                    "InsufficientValidTo",
+                    "validTo is not far enough in the future",
+                ),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::TransferEthToContract => (
+                super::error(
+                    "TransferEthToContract",
+                    "Sending Ether to smart contract wallets is currently not supported",
+                ),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::SameBuyAndSellToken => (
+                super::error(
+                    "SameBuyAndSellToken",
+                    "Buy token is the same as the sell token.",
+                ),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::Other(_) => (super::internal_error(), StatusCode::INTERNAL_SERVER_ERROR),
+        }
+    }
+}
+
+pub struct PreOrderValidator {
+    code_fetcher: Box<dyn CodeFetching>,
+    native_token: WETH9,
+    banned_users: Vec<H160>,
+    min_order_validity_period: Duration,
+}
+
+#[derive(Default)]
+pub struct PreOrderData {
+    owner: H160,
+    sell_token: H160,
+    buy_token: H160,
+    receiver: H160,
+    valid_to: u32,
+    buy_token_balance: BuyTokenDestination,
+    sell_token_balance: SellTokenSource,
```

Comment:
Probably want all of these `pub` too.

Reply:
Yea good call, I made these all public in the next PR, but I can do it here in favour of less changes in the next one.